### PR TITLE
Support remote kernels on web extension

### LINKED ISSUE #9663: Remote jupyter kernel creatable in web extension
Support the 'selectJupyterUri' command and be able to enumerate kernels and connect to them in the web extension

### PR DESCRIPTION
**Pull request by rchiodo**
Fixes #9663 

Refactoring to support web using remote kernels. Essentially you can pick a remote server with the 'Specify server for remote Juptyer connections' and then run kernels against it.

Also refactored some of our test code to be shareable between web extension and the node extension (just remote for now).

### CODE REVIEW

**Comment by DonJayamanne** [comment]
can't see any tests running against this PR

File: src/kernels/jupyter/launcher/notebookServerProvider.ts [hunk 1]
```diff
@@ -176,13 +180,15 @@ export class NotebookServerProvider implements IJupyterServerProvider {
...
-                    : activeInterpreter.uri.fsPath;
+                    : getFilePath(activeInterpreter.uri);
                 throw new Error(
                     DataScience.jupyterNotSupportedBecauseOfEnvironment().format(displayName, e.toString())
                 );
             } else {
                 throw new JupyterInstallError(
-                    DataScience.jupyterNotSupported().format(await this.jupyterExecution.getNotebookError())
+                    DataScience.jupyterNotSupported().format(
+                        this.jupyterExecution ? await this.jupyterExecution.getNotebookError() : 'Error'
```
Comment (IanMatthewHuff) [inline src/kernels/jupyter/launcher/notebookServerProvider.ts:190]: Minor, but I'm pretty sure this is user visible so the 'Error' needs loc.

File: src/kernels/jupyter/remoteKernelFinder.ts [hunk 2]
```diff
@@ -139,8 +143,11 @@ export class RemoteKernelFinder implements IRemoteKernelFinder {
     }
 
     private async getInterpreter(spec: IJupyterKernelSpec, baseUrl: string) {
-        const parsed = new url.URL(baseUrl);
-        if (parsed.hostname.toLocaleLowerCase() === 'localhost' || parsed.hostname === '127.0.0.1') {
+        const parsed = new URL(baseUrl);
+        if (
+            (parsed.hostname.toLocaleLowerCase() === 'localhost' || parsed.hostname === '127.0.0.1') &&
```
Comment (IanMatthewHuff) [inline src/kernels/jupyter/remoteKernelFinder.ts:148]: Do we need something more general here? Like so for edge cases? https://www.npmjs.com/package/is-localhost-ip

File: src/kernels/jupyter/session/jupyterRequestCreator.web.ts [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+                    co_headers = co_headers ? { ...co_headers, ...authorizationHeader } : authorizationHeader;
+                }
+                if (co_headers) {
+                    co = { ...co, headers: co_headers };
+                    console.log(`CO Headers: ${co}`);
+                }
+
+                super(url, protocols);
+
+                // TODO: How to send auth headers? Try debugging this to see what happens.
```
Comment (IanMatthewHuff) [inline src/kernels/jupyter/session/jupyterRequestCreator.web.ts:74]: Old TODO? Looks like you are handling the headers above.

**Review by IanMatthewHuff — APPROVED** [review]
Looks good with a few minor comments. Nice work! 🍾

**Comment by rchiodo** [comment]
> can't see any tests running against this PR

Yeah I messed up the build-test.yml somehow. Need to fix that. I did add a bunch of tests though.

File: src/kernels/jupyter/session/jupyterRequestCreator.web.ts (hunk 3 above)
Comment (rchiodo) [reply src/kernels/jupyter/session/jupyterRequestCreator.web.ts:74]: No this is still a todo. The constructor for the node websocket takes the CO headers. The constructor for the browser websocket doesn't. 

I _think_ they can be applied to each request during the request constructor, but not sure yet.

File: src/kernels/jupyter/remoteKernelFinder.ts (hunk 2 above)
Comment (rchiodo) [reply src/kernels/jupyter/remoteKernelFinder.ts:148]: Oh probably. Separate issue though. This is just preventing it using the interpreter service when there's no python extension.

Comment (rchiodo) [reply src/kernels/jupyter/remoteKernelFinder.ts:148]: Created a new issue #9748

File: src/kernels/jupyter/launcher/notebookServerProvider.ts (hunk 1 above)
Comment (rchiodo) [reply src/kernels/jupyter/launcher/notebookServerProvider.ts:190]: This case will never happen. It's just for the compiler. This is for 'starting' jupyter which we'll never do in web.

**Pull request merged**